aws/resource_aws_neptune_cluster_parameter_group.go

```diff
@@ -0,0 +1,263 @@
+			"family": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"description": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+				Default:  "Managed by Terraform",
+			},
+			"parameter": {
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"name": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"value": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"apply_method": {
+							Type:     schema.TypeString,
+							Optional: true,
+							Default:  neptune.ApplyMethodPendingReboot,
+							ValidateFunc: validation.StringInSlice([]string{
+								neptune.ApplyMethodImmediate,
+								neptune.ApplyMethodPendingReboot,
+							}, false),
+						},
+					},
+				},
+			},
+
+			"tags": tagsSchema(),
+		},
+	}
+}
+
+func resourceAwsNeptuneClusterParameterGroupCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).neptuneconn
+	tags := tagsFromMapNeptune(d.Get("tags").(map[string]interface{}))
+
+	var groupName string
+	if v, ok := d.GetOk("name"); ok {
+		groupName = v.(string)
+	} else if v, ok := d.GetOk("name_prefix"); ok {
+		groupName = resource.PrefixedUniqueId(v.(string))
+	} else {
+		groupName = resource.UniqueId()
+	}
+
+	createOpts := neptune.CreateDBClusterParameterGroupInput{
+		DBClusterParameterGroupName: aws.String(groupName),
+		DBParameterGroupFamily:      aws.String(d.Get("family").(string)),
+		Description:                 aws.String(d.Get("description").(string)),
+		Tags:                        tags,
+	}
+
+	log.Printf("[DEBUG] Create Neptune Cluster Parameter Group: %#v", createOpts)
+	_, err := conn.CreateDBClusterParameterGroup(&createOpts)
+	if err != nil {
+		return fmt.Errorf("Error creating Neptune Cluster Parameter Group: %s", err)
+	}
+
+	d.SetId(aws.StringValue(createOpts.DBClusterParameterGroupName))
+	log.Printf("[INFO] Neptune Cluster Parameter Group ID: %s", d.Id())
+
+	return resourceAwsNeptuneClusterParameterGroupUpdate(d, meta)
+}
+
+func resourceAwsNeptuneClusterParameterGroupRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).neptuneconn
+
+	describeOpts := neptune.DescribeDBClusterParameterGroupsInput{
+		DBClusterParameterGroupName: aws.String(d.Id()),
+	}
+
+	describeResp, err := conn.DescribeDBClusterParameterGroups(&describeOpts)
+	if err != nil {
+		if isAWSErr(err, neptune.ErrCodeDBParameterGroupNotFoundFault, "") {
+			log.Printf("[WARN] Neptune Cluster Parameter Group (%s) not found, removing from state", d.Id())
+			d.SetId("")
+			return nil
+		}
+
+		return err
+	}
+
+	if len(describeResp.DBClusterParameterGroups) != 1 ||
+		aws.StringValue(describeResp.DBClusterParameterGroups[0].DBClusterParameterGroupName) != d.Id() {
+		return fmt.Errorf("Unable to find Cluster Parameter Group: %#v", describeResp.DBClusterParameterGroups)
```

Comment:
Just like the regular parameter groups we should presume the API will return the correct result (terraform's plan would look very wrong otherwise) -- checking for zero length to recreate the resource instead of returning an error:

```go
if len(describeResp.DbClusterParameterGroups) == 0 {
  log.Printf("[WARN] Neptune Cluster Parameter Group (%s) not found, removing from state", d.Id())
  d.SetId("")
  return nil
}
```

Reply:
sure, I understood. But anyway we don't need to check for zero length right? since it should be covered by `ErrCodeDBClusterParameterGroupNotFoundFault`.

Reply:
We've seen cases where some service APIs will respond with no errors and also an empty result set. In any case, its best to include the handling to prevent Terraform from crashing with the Go panic of `index not in range`.